src/home_robot/home_robot/mapping/voxel.py

```diff
@@ -85,17 +231,95 @@ def write_to_pickle(self, filename: str):
             pickle.dump(data, f)
 
     def recompute_map(self):
-        """Recompute the entire map from scratch instead of doing incremental updates"""
-        self.xyz, self.feats = None, None
-        for camera_pose, xyz, feats, _ in self.observations:
+        """Recompute the entire map from scratch instead of doing incremental updates.
+        This is a helper function which recomputes everything from the beginning and migh"""
+        self.reset_cache()
+        for camera_pose, xyz, feats, base_pose, obs, _ in self.observations:
+            # TODO: logic should be the same as in "add" above
+            # TODO: switch to using just Obs struct?
             world_xyz = trimesh.transform_points(xyz, camera_pose)
             if self.xyz is None:
-                self.xyz, self.feats = world_xyz, feats
+                pc_xyz, pc_rgb = world_xyz, feats
             else:
                 pc_rgb = np.concatenate([self.feats, feats], axis=0)
                 pc_xyz = np.concatenate([self.xyz, world_xyz], axis=0)
-        pcd = numpy_to_pcd(pc_xyz, pc_rgb).voxel_down_sample(voxel_size=self.resolution)
-        self.xyz, self.feats = pcd_to_numpy(pcd)
+            if base_pose is not None:
+                self._update_visited(base_pose)
+        self._pcd = numpy_to_pcd(pc_xyz, pc_rgb).voxel_down_sample(
+            voxel_size=self.resolution
+        )
+        self.xyz, self.feats = pcd_to_numpy(self._pcd)
+
+    def get_2d_map(
+        self, debug: bool = False
+    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+        """Get 2d map with explored area and frontiers."""
+
+        # Is this already cached? If so we don't need to go to all this work
+        if self._map2d is not None and self._seq == self._2d_last_updated:
+            return self._map2d
+
+        # Convert metric measurements to discrete
+        # Gets the xyz correctly - for now everyhting is assumed to be within correct distance of origin
+        xyz = ((self.xyz / self.grid_resolution) + self.grid_origin).astype(np.uint32)
+
+        # Crop to robot height
+        min_height = int(self.obs_min_height / self.grid_resolution)
+        max_height = int(self.obs_max_height / self.grid_resolution)
+        # NOTE: keep this if we only care about obstacles
+        # voxels = np.zeros(self.grid_size + [int(max_height - min_height)])
+        # But we might want to track floor pixels as well
+        voxels = np.zeros(self.grid_size + [max_height])
+        # NOTE: you can use min_height for this if we only care about obstacles
+        obs_mask = np.bitwise_and(xyz[:, -1] > 0, xyz[:, -1] < max_height)
+        obs_mask = xyz[:, -1] < max_height
```

Comment:
The second line overrides what is written in the first line

Reply:
yep, first line is old code (doesnt track floor space)